src/otg/preprocess/finngen/study_index.py

```diff
@@ -0,0 +1,128 @@
+"""FinnGen ingestion: study index."""
+
+from __future__ import annotations
+
+import logging
+import sys
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+from urllib.request import urlopen
+
+import pyspark.sql.functions as f
+from pyspark.sql import SparkSession
+
+from otg.common.schemas import parse_spark_schema
+from otg.dataset.study_index import StudyIndex
+
+if TYPE_CHECKING:
+    from pyspark.sql import DataFrame
+    from pyspark.sql.types import StructType
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s %(message)s")
+
+
+@dataclass
+class FinnGenStudyIndex(StudyIndex):
+    """Study index dataset from FinnGen.
+
+    The following information is aggregated/extracted:
+
+    - Study ID in the special format (FINNGEN_R9_*)
+    - Trait name (for example, Amoebiasis)
+    - Number of cases and controls
+    - Link to the summary statistics location
+
+    Some fields are also populated as constants, such as study type and the initial sample size.
+    """
+
+    @classmethod
+    def get_schema(cls: type[FinnGenStudyIndex]) -> StructType:
+        """Provides the schema for the FinnGenStudyIndex dataset."""
+        return parse_spark_schema("study_index.json")
+
+    @classmethod
+    def from_source(
+        cls: type[FinnGenStudyIndex],
+        finngen_studies: DataFrame,
+        finngen_release_prefix: str,
+        finngen_summary_stats_url_prefix: str,
+        finngen_summary_stats_url_suffix: str,
+    ) -> FinnGenStudyIndex:
+        """This function ingests study level metadata from FinnGen.
+
+        Args:
+            finngen_studies (DataFrame): FinnGen raw study table
+            finngen_release_prefix (str): Release prefix pattern.
+            finngen_summary_stats_df_url_prefix (str): URL prefix for summary statistics location.
+            finngen_summary_stats_df_url_suffix (str): URL prefix suffix for summary statistics location.
+
+        Returns:
+            FinnGenStudyIndex: Parsed and annotated FinnGen study table.
+        """
+        return cls(
+            _df=finngen_studies.select(
+                f.concat(f.lit(f"{finngen_release_prefix}_"), f.col("phenocode")).alias(
+                    "studyId"
+                ),
+                f.col("phenostring").alias("traitFromSource"),
+                f.col("num_cases").alias("nCases"),
+                f.col("num_controls").alias("nControls"),
+                (f.col("num_cases") + f.col("num_controls")).alias("nSamples"),
+                f.lit(finngen_release_prefix).alias("projectId"),
+                f.lit("gwas").alias("studyType"),
+                f.lit(True).alias("hasSumstats"),
+                f.lit("377,277 (210,870 females and 166,407 males)").alias(
```

Comment:
Some of the recently added logic from main is missing here:
- Adding column ["discoverySamples"](https://github.com/opentargets/genetics_etl_python/blob/f60e7205ea1c4306b85099e9fcb4e149b3ec5e1d/src/otg/datasource/finngen/study_index.py#L60)
- Adding column ["ldPopulationStructure"](https://github.com/opentargets/genetics_etl_python/blob/f60e7205ea1c4306b85099e9fcb4e149b3ec5e1d/src/otg/datasource/finngen/study_index.py#L76)

Reply:
Thank you for spotting this, I've now reconciled all changes and added those as well.